[USMON-358] postgres monitoring

guyarb opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Add initial and simple support for USM to capture postgres traffic.
USM is now able to capture plaintext simple postgres queries, and report the operation (select, insert, update, drop, create table) and the table name, along with latency and request count.
The PR makes the initial ramp up around adding a new protocol to USM, and intentionally keeps the simplicity of tracked queries.
In a series of following PR will add support to more complicated queries, more postgres scenarios, add postgres-TLS support, and more.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

USM's roadmap is to expand to new protocols, especially DB protocols.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

- [x] Merge agent-payload [PR](https://github.com/DataDog/agent-payload/pull/303)
- [x] Dogfooding
- [x] Add UT suite

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

guyarb committed:
usm: Update agent-payload version

Updating to a version that supports DB aggregations from USM

guyarb committed:
usm: postgres: Added configuration to enable postgres monitoring

guyarb committed:
usm: postgres: Added configuration to control max stats buffered

Every protocol has its own configuration to control the number of stats in memory
to store until the next connections check

guyarb committed:
usm: postgres: Added eBPF decoding code

The commit contains the postgres decoding plaintext implementation.

guyarb committed:
usm: postgres: Added protocol

The commit adds the postgres protocol struct to USM.
The commit does not contain actual user-mode processing of postgres events. That part will be
added in a dedicated commit

guyarb committed:
usm: postgres: Add test build flag, to test utils files

guyarb committed:
usm: postgres: Add statskeeper and processing of events

guyarb committed:
usm: postgres: Expose postgres requests to connections check

guyarb committed:
usm: postgres: Added debugging endpoint

val06 reviewed (approved):
LGTM (ebpf-platform files)

pgimalac reviewed (approved):
LGTM for ASC owned files

amitslavin commented on pkg/config/setup/system_probe.go:
```diff
@@ -246,6 +247,7 @@ func InitSystemProbeConfig(cfg pkgconfigmodel.Config) {
 	cfg.BindEnv(join(netNS, "max_http_stats_buffered"), "DD_SYSTEM_PROBE_NETWORK_MAX_HTTP_STATS_BUFFERED")
 	cfg.BindEnv(join(smNS, "max_http_stats_buffered"))
 	cfg.BindEnvAndSetDefault(join(smNS, "max_kafka_stats_buffered"), 100000)
+	cfg.BindEnv(join(smNS, "max_postgres_stats_buffered"))
```
Same question as above for here, I saw that you used the cmd/system-probe/config/adjust_usm.go to set the value. But I am wondering what is the correct way to do so, currently seems like we do both

guyarb replied:
same as https://github.com/DataDog/datadog-agent/pull/25416#discussion_r1592409582

amitslavin commented on pkg/network/ebpf/c/protocols/postgres/decoding.h:
```diff
@@ -0,0 +1,126 @@
+        return 0;
+    }
+
+    normalize_tuple(&conn_tuple);
```
Do we still need it?

guyarb replied:
yes, still relevant for plaintext

amitslavin commented on pkg/network/ebpf/c/protocols/postgres/types.h:
```diff
@@ -0,0 +1,31 @@
+#define POSTGRES_BATCH_SIZE 25
+
+// Maximum length of postgres query to send to userspace.
+#define POSTGRES_BUFFER_SIZE 64
+
+// Maximum number of Postgres messages we can parse for a single packet.
+#define POSTGRES_MAX_MESSAGES 100
```
Do we have any insight if those values are set correctly by staging or production traffic?

guyarb replied:
no, that's a chicken and egg problem
one of the followup tasks will be to measure the assumptions I took

amitslavin commented on pkg/network/ebpf/c/protocols/postgres/types.h:
```diff
@@ -0,0 +1,31 @@
+    __u64 request_started;
+    __u64 response_last_seen;
+    // The actual size of the query stored in request_fragment.
+    __u32 original_query_size;
```
why do we need original_query_size? in order to handle the buffer in the user mode?

guyarb replied:
> in order to handle the buffer in the user mode?
true
I also think it will allow us to know if the query was truncated (original_query_size > 64) if case it will matter to us

amitslavin commented on pkg/network/protocols/postgres/debugging/debugging.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+func formatIP(low, high uint64) util.Address {
+	// for Kafka at the moment, so given this is purely debugging code I think it's fine
+	// to assume for now that it's only IPv6 if higher order bits are set.
```
I do not think it is needed here

amitslavin commented on pkg/network/protocols/postgres/operations.go:
```diff
@@ -0,0 +1,56 @@
+type Operation uint8
+
+const (
+	UnknownOP Operation = iota
```
Fix import issue

amitslavin commented on pkg/network/protocols/postgres/protocol.go:
```diff
@@ -0,0 +1,185 @@
+
+	// Clean up idle connections. We currently use the same TTL as HTTP, but we plan to rename this variable to be more generic.
+	ttl := p.cfg.HTTPIdleConnectionTTL.Nanoseconds()
+	mapCleaner.Clean(p.cfg.HTTPMapCleanerInterval, nil, nil, func(now int64, key netebpf.ConnTuple, val EbpfTx) bool {
```
This is a bit out of scope, but I think we should change this variable name. We are using this interval time in different protocols, and it is always based on 'HTTP'. That seems odd to me. If you agree, I can take care of it

guyarb replied:
true

amitslavin commented on pkg/network/protocols/postgres/stats_structs.go:
```diff
@@ -0,0 +1,67 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package postgres
```
Some CI error to handle

amitslavin commented on pkg/network/state.go:
```diff
@@ -277,7 +285,7 @@ type networkState struct {
 
 // NewState creates a new network state
-func NewState(clientExpiry time.Duration, maxClosedConns uint32, maxClientStats int, maxDNSStats int, maxHTTPStats int, maxKafkaStats int, enableConnectionRollup bool, processEventConsumerEnabled bool) State {
+func NewState(clientExpiry time.Duration, maxClosedConns uint32, maxClientStats, maxDNSStats, maxHTTPStats, maxKafkaStats, maxPostgresStats int, enableConnectionRollup bool, processEventConsumerEnabled bool) State {
```
You need to fix the callers of this function to support your changes.

leeavital reviewed (approved):
Light review for NPM owned files

amitslavin commented on pkg/network/ebpf/c/protocols/postgres/decoding.h:
```diff
@@ -0,0 +1,126 @@
+    // offset to the end of the message header, we want to jump over the payload.
+    skb_info.data_off += header.message_len - sizeof(__u32);
+
+#pragma unroll(POSTGRES_MAX_MESSAGES)
```
Lets add to the documentation some links to the RFC, Here can be a nice place to link the documnation for the command complete message, and the option to response with multiple responses.

As well let's link the message format document - https://www.postgresql.org/docs/current/protocol-message-formats.html somewhere, maybe here before the handle_new_query function

guyarb commented on pkg/config/setup/system_probe.go:
```diff
@@ -231,6 +231,7 @@ func InitSystemProbeConfig(cfg pkgconfigmodel.Config) {
 
 	cfg.BindEnvAndSetDefault(join(smNS, "enable_http2_monitoring"), false)
 	cfg.BindEnvAndSetDefault(join(smNS, "enable_kafka_monitoring"), false)
+	cfg.BindEnv(join(smNS, "enable_postgres_monitoring"))
```
By default both will be false, but if we use `SetDefault` we wouldn't be able to know if the user `set` the value or we did
The usage of http2 and kafka is wrong but out of scope for that PR

amitslavin replied:
Do you want me to change it for HTTP/2 and create tasks for both?

AmitaiBl commented on cmd/system-probe/config/adjust_usm.go:
```diff
@@ -47,6 +47,7 @@ func adjustUSM(cfg config.Config) {
 	applyDefault(cfg, spNS("process_service_inference", "enabled"), false)
 	applyDefault(cfg, spNS("process_service_inference", "use_windows_service_name"), true)
 	applyDefault(cfg, smNS("enable_ring_buffers"), true)
+	applyDefault(cfg, smNS("max_postgres_stats_buffered"), 100000)
```
can this value be in a config file?

guyarb replied:
yes, 
```yaml
service_moniotring_config:
  max_postgres_stats_buffered: 15
```

AmitaiBl commented on pkg/network/protocols/postgres/model_linux.go:
```diff
@@ -0,0 +1,116 @@
+	regexMapper = map[Operation]*regexp.Regexp{
+		CreateTableOP: regexp.MustCompile(`CREATE TABLE (\S+)`),
+		InsertOP:      regexp.MustCompile(`INSERT INTO (\S+)`),
+		DropTableOP:   regexp.MustCompile(`DROP TABLE (\S+)`),
```
only tracking drop of tables? what about drop Index?

guyarb replied:
Generally we're going to focus on TABLE operations, so schema, index, tableops, and others will be out of scope of USM

AmitaiBl commented on pkg/network/protocols/postgres/model_linux.go:
```diff
@@ -0,0 +1,116 @@
+	return e.Request_fragment[:e.Original_query_size]
+}
+
+// Operation returns the operation of the query (SELECT, INSERT, UPDATE, DROP, etc.)
```
what about `ALTER` statement, are this not meant to be tracked?

guyarb replied:
https://github.com/DataDog/datadog-agent/pull/25416#discussion_r1592535282

AmitaiBl commented on pkg/network/protocols/postgres/model_linux.go:
```diff
@@ -0,0 +1,116 @@
+		if matches := regex.FindSubmatch(e.Tx.getFragment()); len(matches) > 1 {
+			return string(
+				bytes.Trim( // Remove leading and trailing quotes from the table name
+					bytes.ReplaceAll(matches[1], []byte{0}, []byte{}), // Remove null bytes
```
what about the case that create a table from a select statement, will you now have 2 matches here?

guyarb replied:
We'll take the first match `matches[1]`
For the initial version that's sufficient, we're looking to capture simple queries and leaving the multi-statement, or other types of queries aside

AmitaiBl commented on pkg/network/protocols/postgres/operations.go:
```diff
@@ -0,0 +1,56 @@
+	case UpdateOP:
+		return "UPDATE"
+	case CreateTableOP:
+		return "CREATE"
```
what about `ALTER` operation

guyarb replied:
We intentionally kept the PR very simple on the types of queries we're processing

> The PR makes the initial ramp up around adding a new protocol to USM, and intentionally keeps the simplicity of tracked queries.
In a series of following PR will add support to more complicated queries, more postgres scenarios, add postgres-TLS support, and more.

Since the PR is complex and touches many files we do the ramping up here, and with a followup pr, and decided ones, we'll add support for more use cases and more queries types

guyarb committed:
Fixed CR comments

guyarb committed:
usm: postgres: Added postgres encoder

guyarb committed:
usm: postgres: Added UTs

guyarb committed:
usm: postgres: Handle verifier error on old kernels

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+	targetAddress := net.JoinHostPort(targetHost, postgresPort)
+	require.NoError(t, postgres.RunServer(t, serverHost, postgresPort))
+
+	//defaultDialer := &net.Dialer{
+	//	LocalAddr: &net.TCPAddr{
+	//		IP: net.ParseIP(clientHost),
+	//	},
+	//}
```
remove

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+func (s *postgresProtocolParsingSuite) TestDecoding() {
+	t := s.T()
+
+	//clientHost := "localhost"
```
remove

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+	t := s.T()
+
+	//clientHost := "localhost"
+	targetHost := "127.0.0.1"
+	serverHost := "127.0.0.1"
```
Does the target and servers hosts are equal in purpose?

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+	postgres.ConnectAndGetDB(t, ctx.serverAddress, ctx.extras)
+	postgres.RunCreateQuery(t, ctx.extras)
+	values := make([]string, 5)
+	for i := 0; i < 20; i++ {
```
why 20? 
add documentation

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+	}
+}
+
+func validatePostgres(t *testing.T, monitor *Monitor, expectedStats map[string]map[postgres.Operation]int) {
```
Perhaps we should organize all the tests and functions in the same order?

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+			},
+		},
+		{
+			name: "insert",
```
Perhaps we could give the test a more descriptive name that explains its purpose?

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+				postgres.RunCreateQuery(t, ctx.extras)
+			},
+			postMonitorSetup: func(t *testing.T, ctx testContext) {
+				for i := 0; i < 2; i++ {
```
why 2? 
add doc

amitslavin commented on pkg/network/usm/postgres_monitor_test.go:
```diff
@@ -0,0 +1,313 @@
+			preMonitorSetup: func(t *testing.T, ctx testContext) {
+				postgres.ConnectAndGetDB(t, ctx.serverAddress, ctx.extras)
+				postgres.RunCreateQuery(t, ctx.extras)
+				for i := 0; i < 20; i++ {
```
same for here as above

guyarb replied:
removed

amitslavin commented on pkg/network/encoding/marshal/postgres_test.go:
```diff
@@ -0,0 +1,313 @@
+func (s *PostgresSuite) TestFormatPostgresStats() {
+	t := s.T()
+
+	postgresKey1 := postgres.NewKey(
+		localhost,
+		localhost,
+		postgresClientPort,
+		postgresServerPort,
+		postgres.SelectOP,
+		tableName,
+	)
+	postgresKey2 := postgres.NewKey(
+		localhost,
+		localhost,
+		postgresClientPort,
+		postgresServerPort,
+		postgres.InsertOP,
+		tableName,
+	)
```
I believe we can use better naming as `selectKey` and `insertKey`

guyarb committed:
usm: postgres: Fixed wrong latnecy for multiple records

If we have a single record, then we should use firstLatencySample, but if we have
many records, we should use the latency field

guyarb committed:
usm: postgres: Update protobuf structs

guyarb committed:
usm: postgres: tests: Address CR comments

guyarb commented:
/merge

derekwbrown reviewed (approved):
approved for (2) files owned by wkit

guyarb commented:
/merge
